Bundle Analysis Tool

narasimha-1511 opened the pull request:
## Changes
- Implemented robust transaction history fetching using Helius RPC methods
- Added `getSignaturesForAddress` and `getTransaction` to fetch detailed transaction data

I had taken all the transactions i grouped them 

calculated 
 purchase velocity : tokens purchased per hour
 checked if has high velocity or  has supply percentage > 1 
 checked if his purchases were quick ? and also if he is a early buyer ?
. 
 This pull request was created for https://app.gib.work/bounties/a2598bfe-736d-4623-87db-f2df00e1a2db in an attempt to solve a bounty #34 . Payment for the bounty is immediately sent to the contributor after merge.

narasimha-1511 commented:
Hey @slimeonmyhead  can you give some feed back or any changes needed ??

slimeonmyhead commented:
@narasimha-1511 can you show a working example of this? I'm getting rate limited on our helius API key if I try to get the transactions of the mint.

At the very least we can batch the transaction data call:
```
const txs = await connection.getSignaturesForAddress(address);
const graph = new Map();
const chunks = chunk(txs, 25);

for (const batch of chunks) {
  const details = await connection.getTransactions(
    batch.map(tx => tx.signature),
    { maxSupportedTransactionVersion: 0 }
  );
  // Process each transaction in the batch
  batch.forEach((tx, i) => {
    if (details[i]) {
      graph.set(tx.signature, details[i]);
    }
  });
}
```


Right now, `getTransactionHistory` would only get the most recent 100 transactions. We're more interested in the first transactions on the mint address

slimeonmyhead commented on src/ai/solana/bundle.tsx:
```diff
@@ -0,0 +1,236 @@
+    description:
+      'Analyze potential bundles and snipers for a given mint address, including statistics about supply percentage, estimated SOL spent, and current holdings.',
+    parameters: z.object({
+      mintAddress: z.string().describe("The NFT collection's mint address"),
```
Should be
```
mintAddress: z.string().describe("The token's mint address"),
```

narasimha-1511 replied:
resolved

slimeonmyhead commented on src/lib/solana/helius.ts:
```diff
@@ -401,3 +408,40 @@ export async function getHoldersClassification(
+  // Then get full transaction data for each signature
+  const transactions = await Promise.all(
+    signaturesData.result.map(async (sig: any) => {
+      const txData = await fetchHelius('getTransaction', [sig.signature]);
```
Need to pass in the `maxSupportedTransactionVersion` parameter

```
const txData = await fetchHelius('getTransaction', [
  sig.signature,
  { maxSupportedTransactionVersion: 0 },
]);
```

narasimha-1511 replied:
resolved this

narasimha-1511 commented:
Hey @slimeonmyhead 

## Changes
- Implemented rate limit handling for Helius API requests
- Added buyer tracking in transaction analysis
- Enhanced bundle detection with unique wallet tracking
- Optimized transaction processing with batch requests


small video of the response i got , 
i was unable to  chat , i was getting openai error


https://github.com/user-attachments/assets/6d0e51f2-f245-414a-a769-7c81267da200

narasimha-1511 commented:
Hey @slimeonmyhead 

Changes - 
 - Getting the transactions in bacthes of 100
 - used pagination in getting the signatures
 - here is a small chat how it worked
 
![Uploading image.png…]()

narasimha-1511 commented:
@slimeonmyhead 

 Changes Made

- Update bundle analysis layout
- Add detailed bundle metrics display with wallet breakdowns
- Integrate with ne wAPI response structure
- Improve error handling and type safety
- Add bundle numbering and better visual hierarchy
- Added a show more option for more than three bundles

diffrence between the regular and snipers addresses

![image](https://github.com/user-attachments/assets/e6dfc45e-39ed-4b4f-8890-70159aec2314)
![image](https://github.com/user-attachments/assets/e9cbb4c2-6197-41e0-ab34-bf00cf0810fb)


A small video of the redesigned UI

https://github.com/user-attachments/assets/61b5d5c5-6f1b-4929-8327-20726cb9c324

narasimha-1511 commented:
changes made

- added next/previous 


https://github.com/user-attachments/assets/8c82bcc7-3853-4d4a-8a72-a32943eac01d
